### ISSUE TITLE
AttributeError: 'Model' object has no attribute 'loss_functions'

### ISSUE THREAD

**Issue by kangPrayit** [amaiya/ktrain#112]
I am using
Ktrain version :  0.12.3
Tensorflow Version:  2.2.0-rc2

then try the tutorial on ['Explaining prediction and Misclassification'](https://nbviewer.jupyter.org/github/amaiya/ktrain/blob/master/tutorials/tutorial-A2-explaining-predictions.ipynb)

I always get this error: 
`/usr/local/lib/python3.6/dist-packages/ktrain/core.py in top_losses(self, n, val_data, preproc)
    193         #losses = self.model.loss_functions[0](tf.convert_to_tensor(y_true), tf.convert_to_tensor(y_pred))
    194         if U.is_tf_keras():
--> 195             L = self.model.loss_functions[0].fn
    196         else:
    197             L = self.model.loss_functions[0]
AttributeError: 'Model' object has no attribute 'loss_functions'`

help really appreciate it.

Thank you.

**Comment by amaiya** [comment]
Thanks for letting me know about this.  

I've confirmed that everything works in TensorFlow 2.1.0, which is the current/stable release of TensorFlow.  (TF 2.2.0 is not officially released.)  

This appears to be an undocumented breaking change, as `model.loss_functions` does not exist in TensorFlow 2.2.0-rc2.  

More specifically, in TensorFlow 2.2.0, the loss function(s) associated with the model are more buried and accessed like this: ```learner.model.compiled_loss._losses[0].fn``` instead of this ```learner.model.loss_functions[0].fn```.

For now, it will work if you downgrade to TensorFlow to 2.1.0, which is the latest stable release.  

When TensorFlow 2.2.0 is officially released, we will also release a new version of *ktrain* to ensure *ktrain* `view_top_losses` continues to work with TF 2.2.0.

Thanks again.

**Comment by nikolamilosevic86** [comment]
I am not sure whether this is associated issue, but I am getting the following error, trying to train with albert
```
simulating training for different learning rates... this may take a few moments...
preparing train data ...done.
Train for 4911 steps
Epoch 1/1024
   0/4911 [..............................] - ETA: 0s

---------------------------------------------------------------------------
InvalidArgumentError                      Traceback (most recent call last)
~/.local/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_v2.py in on_batch(self, step, mode, size)
    783       try:
--> 784         yield batch_logs
    785       finally:

~/.local/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_v2.py in run_one_epoch(model, iterator, execution_function, dataset_size, batch_size, strategy, steps_per_epoch, num_samples, mode, training_context, total_epochs)
    127       try:
--> 128         batch_outs = execution_function(iterator)
    129       except (StopIteration, errors.OutOfRangeError):

~/.local/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_v2_utils.py in execution_function(input_fn)
     97     return nest.map_structure(_non_none_constant_value,
---> 98                               distributed_function(input_fn))
     99 

~/.local/lib/python3.6/site-packages/tensorflow_core/python/eager/def_function.py in __call__(self, *args, **kwds)
    567     else:
--> 568       result = self._call(*args, **kwds)
    569 

~/.local/lib/python3.6/site-packages/tensorflow_core/python/eager/def_function.py in _call(self, *args, **kwds)
    631         # stateless function.
--> 632         return self._stateless_fn(*args, **kwds)
    633     else:

~/.local/lib/python3.6/site-packages/tensorflow_core/python/eager/function.py in __call__(self, *args, **kwargs)
   2362       graph_function, args, kwargs = self._maybe_define_function(args, kwargs)
-> 2363     return graph_function._filtered_call(args, kwargs)  # pylint: disable=protected-access
   2364 

~/.local/lib/python3.6/site-packages/tensorflow_core/python/eager/function.py in _filtered_call(self, args, kwargs)
   1610                            resource_variable_ops.BaseResourceVariable))),
-> 1611         self.captured_inputs)
   1612 

~/.local/lib/python3.6/site-packages/tensorflow_core/python/eager/function.py in _call_flat(self, args, captured_inputs, cancellation_manager)
   1691       return self._build_call_outputs(self._inference_function.call(
-> 1692           ctx, args, cancellation_manager=cancellation_manager))
   1693     forward_backward = self._select_forward_and_backward_functions(

~/.local/lib/python3.6/site-packages/tensorflow_core/python/eager/function.py in call(self, ctx, args, cancellation_manager)
    544               attrs=("executor_type", executor_type, "config_proto", config),
--> 545               ctx=ctx)
    546         else:

~/.local/lib/python3.6/site-packages/tensorflow_core/python/eager/execute.py in quick_execute(op_name, num_outputs, inputs, attrs, ctx, name)
     66       message = e.message
---> 67     six.raise_from(core._status_to_exception(e.code, message), None)
     68   except TypeError as e:

~/.local/lib/python3.6/site-packages/six.py in raise_from(value, from_value)

InvalidArgumentError: 2 root error(s) found.
  (0) Invalid argument:  ConcatOp : Dimensions of inputs should match: shape[0] = [32,70,300] vs. shape[1] = [32,88,3072]
	 [[node model_1/concatenate_1/concat (defined at /home/ubuntu/.local/lib/python3.6/site-packages/ktrain/lroptimize/lrfinder.py:117) ]]
	 [[loss/dense_3_loss/weighted_loss/broadcast_weights/assert_broadcastable/AssertGuard/pivot_f/_15/_63]]
  (1) Invalid argument:  ConcatOp : Dimensions of inputs should match: shape[0] = [32,70,300] vs. shape[1] = [32,88,3072]
	 [[node model_1/concatenate_1/concat (defined at /home/ubuntu/.local/lib/python3.6/site-packages/ktrain/lroptimize/lrfinder.py:117) ]]
0 successful operations.
0 derived errors ignored. [Op:__inference_distributed_function_107854494]

Function call stack:
distributed_function -> distributed_function


During handling of the above exception, another exception occurred:

KeyError                                  Traceback (most recent call last)
<ipython-input-40-f01cf5c6afa7> in <module>()
----> 1 learner.lr_find()

~/.local/lib/python3.6/site-packages/ktrain/core.py in lr_find(self, start_lr, lr_mult, max_epochs, stop_factor, show_plot, verbose)
    453                                 use_multiprocessing=self.use_multiprocessing,
    454                                 batch_size=self.batch_size,
--> 455                                 verbose=verbose)
    456         except KeyboardInterrupt:
    457             # re-load current weights

~/.local/lib/python3.6/site-packages/ktrain/lroptimize/lrfinder.py in find(self, train_data, steps_per_epoch, use_gen, start_lr, lr_mult, max_epochs, batch_size, workers, use_multiprocessing, verbose)
    115                    workers=workers, use_multiprocessing=use_multiprocessing,
    116                    verbose=verbose,
--> 117                    callbacks=[callback])
    118         else:
    119             self.model.fit(train_data[0], train_data[1],

~/.local/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training.py in fit(self, x, y, batch_size, epochs, verbose, callbacks, validation_split, validation_data, shuffle, class_weight, sample_weight, initial_epoch, steps_per_epoch, validation_steps, validation_freq, max_queue_size, workers, use_multiprocessing, **kwargs)
    817         max_queue_size=max_queue_size,
    818         workers=workers,
--> 819         use_multiprocessing=use_multiprocessing)
    820 
    821   def evaluate(self,

~/.local/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_v2.py in fit(self, model, x, y, batch_size, epochs, verbose, callbacks, validation_split, validation_data, shuffle, class_weight, sample_weight, initial_epoch, steps_per_epoch, validation_steps, validation_freq, max_queue_size, workers, use_multiprocessing, **kwargs)
    340                 mode=ModeKeys.TRAIN,
    341                 training_context=training_context,
--> 342                 total_epochs=epochs)
    343             cbks.make_logs(model, epoch_logs, training_result, ModeKeys.TRAIN)
    344 

~/.local/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_v2.py in run_one_epoch(model, iterator, execution_function, dataset_size, batch_size, strategy, steps_per_epoch, num_samples, mode, training_context, total_epochs)
    179             batch_end=step * batch_size + current_batch_size)
    180       cbks.make_logs(model, batch_logs, batch_outs, mode)
--> 181       step += 1
    182 
    183     if callbacks.model.stop_training:

~/anaconda3/envs/tensorflow2_p36/lib/python3.6/contextlib.py in __exit__(self, type, value, traceback)
     97                 value = type()
     98             try:
---> 99                 self.gen.throw(type, value, traceback)
    100             except StopIteration as exc:
    101                 # Suppress StopIteration *unless* it's the same exception that

~/.local/lib/python3.6/site-packages/tensorflow_core/python/keras/engine/training_v2.py in on_batch(self, step, mode, size)
    786         if not batch_logs.pop('data_exhausted', False):
    787           self.callbacks._call_batch_hook(
--> 788               mode, 'end', step, batch_logs)
    789           self.progbar.on_batch_end(step, batch_logs)

~/.local/lib/python3.6/site-packages/tensorflow_core/python/keras/callbacks.py in _call_batch_hook(self, mode, hook, batch, logs)
    237     for callback in self.callbacks:
    238       batch_hook = getattr(callback, hook_name)
--> 239       batch_hook(batch, logs)
    240     self._delta_ts[hook_name].append(time.time() - t_before_callbacks)
    241 

~/.local/lib/python3.6/site-packages/tensorflow_core/python/keras/callbacks.py in on_train_batch_end(self, batch, logs)
    526     """
    527     # For backwards compatibility.
--> 528     self.on_batch_end(batch, logs=logs)
    529 
    530   @doc_controls.for_subclass_implementers

~/.local/lib/python3.6/site-packages/ktrain/lroptimize/lrfinder.py in <lambda>(batch, logs)
    102         K.set_value(self.model.optimizer.lr, start_lr)
    103 
--> 104         callback = LambdaCallback(on_batch_end=lambda batch, logs: self.on_batch_end(batch, logs))
    105 
    106 

~/.local/lib/python3.6/site-packages/ktrain/lroptimize/lrfinder.py in on_batch_end(self, batch, logs)
     28 
     29         # Log the loss
---> 30         loss = logs['loss']
     31         self.batch_num +=1
     32         self.avg_loss = self.beta * self.avg_loss + (1-self.beta) *loss

KeyError: 'loss'


```

**Comment by amaiya** [comment]
@nikolamilosevic86  The issue you described is different.  The issue you posted seems to be with `lr_find` method (which is used to estimate the learning rate).  It did not work in older versions of `transformers` due to a bug with AlBERT models, but works in the latest version of the `transformers` library.  So, you can try upgrading `transformers` to the latest version, if you're using an older version.

In general, learning rates between `1e-5` and `5e-5` tend to work well with Transformer models, so the learning-rate-finder may not be that necessary for your.   The [original AlBERT paper](https://arxiv.org/pdf/1909.11942.pdf) also found something similar, I believe- see the last page in the paper where they show the learning rates they used for different tasks (which are all between 1e-5 and 5e-5).

That said, if trying to use a faster version of BERT, I generally use DistilBERT.

**Comment by nikolamilosevic86** [comment]
Ok cool. Thanks for this answer and suggestions, that is quite helpful. 

Also, it seems that what I have posted was likely error due to running out of memory (even though it reported different thing). I think that when I reduced a bit the size of network and batch size it trained ok (although it was a couple of days ago and I don't remember all the things I did exactly, but I managed to make it train).
